src/System.Runtime/src/System/HashCode.cs

```diff
@@ -239,35 +257,26 @@ private static uint Rol(uint value, int count)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static void Initialize(out uint v1, out uint v2, out uint v3, out uint v4)
         {
-            unchecked
-            {
-                v1 = _seed + Prime1 + Prime2;
-                v2 = _seed + Prime2;
-                v3 = _seed + 0;
-                v4 = _seed - Prime1;
-            }
+            var seed = s_seed;
+            v1 = seed + Prime1 + Prime2;
+            v2 = seed + Prime2;
+            v3 = seed + 0;
+            v4 = seed - Prime1;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint Round(uint hash, uint input)
+        private static void Round(ref uint hash, uint input)
```

Comment:
`ref` is to keep inlining happy.

Reply:
Given the comment I'm not sure whether you've benchmarked that specific case, or you mean this as a general observation. So I'll try to minimize potential damage, risking telling you the obvious:

I've used the same pattern myself, but mostly to more easily compare different hasher implementations. At least from benchmarking results, using ref vs. simple return values does not make a difference on its own, _but_ sometimes making a temporary, local copy of a variable _does_. While this here will be fine, I'd recommend looking at the jitted disassembly and have extensive benchmarks before generalizing the idea that manipulating state via `ref` in/out vars is always advantageous.

Reply:
Check this comment out a bit further up the thread: https://github.com/dotnet/corefx/pull/25013#discussion_r148850063

Basically, writing to an input parameter prevents inlining, according to [this tool](https://marketplace.visualstudio.com/items?itemName=StephanZehetner.InliningAnalyzer) (which invokes RyuJIT and so reflects actual behavior). All of these methods failed for this reason.

Reply:
Thanks for the background info!

Reply:
Good feedback regardless, need to share my benchmarks :). Check the OP for the project.

Reply:
> Basically, writing to an input parameter (callee) prevents inlining, according to this tool (which invokes RyuJIT and so reflects actual behavior). All of these methods failed for this reason.

Storing to a method parameter shouldn't block inlining anymore, this has been fixed in https://github.com/dotnet/coreclr/pull/6572. You might be using an old .NET Framework version if you see that reason.

Reply:
The benchmark results make more sense now - removing the switch statements caused the improvement (explaining why the inlining changes made no difference for the unrolled methods). I'm downloading 4.7 now and will make further changes against that runtime.

Reply:
It turns out the `AggressiveInlining` is needed on everything but MixEmptyState - [Jit Dumps](https://gist.github.com/jcdickinson/cd6885bba371a68a17fb475a491219c9).